core/consensus/beefy/impl/beefy_impl.cpp

```diff
@@ -307,6 +311,15 @@ namespace kagome::network {
       return outcome::success();
     }
     pending_justifications_.emplace(block_number, std::move(justification));
+
+    synchronizer_.get()->fetchHeadersBack(block_tree_->getLastFinalized().number, block_number,
+      [WEAK_SELF] (auto&& res) {
+        WEAK_LOCK(self);
+        if (auto er = self->fetchHeaders(); er.has_error()) {
+          SL_WARN(self->log_, "Failed to fetch headers: {}", er.error().message());
+        }
+      });
+
```

Comment:
Move to `fetchHeaders`

Reply:
block_number is taken from justification.commitment.block_number in onJustification. Without knowing that we can't say what's the minimal block number to save.

Reply:
When `apply` is called first, it means that we have all justifications and blocks to prove beefy chain, so `fetchHeaders` loop is not active.
(so this condition is useless)

But when we don't have sufficient blocks (`fetchHeaders` is called) and don't receive justification (`apply` is not called),
we must request justification,
and for that we need headers.
(so this code is unreachable)